Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -2461,6 +2461,13 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
     if (completedURL.protocolIsJavaScript() && frameElement() && !frameElement()->document().contentSecurityPolicy()->allowJavaScriptURLs(aboutBlankURL().string(), { }, completedURL.string(), frameElement()))
         return;
 
+    RefPtr mainFrame = dynamicDowncast<LocalFrame>(frame->mainFrame());
+    // If the loader for activeWindow's frame (browsing context) has no outgoing referrer, set its outgoing referrer
+    // to the URL of its main frame's Document — which is the incumbent global object's associated Document.
+    // https://html.spec.whatwg.org/#location-object-navigate
+    if (RefPtr activeFrame = activeWindow.frame(); activeFrame && activeFrame->loader().outgoingReferrer().isEmpty() && mainFrame)
+        activeFrame->loader().setOutgoingReferrer(mainFrame->document()->urlForBindings());
```

Comment:
urlForBindings looks a little strange here.  This is not the bindings.  We're also missing a call to strippedForUseAsReferrer.  Also, I think we may want to restrict this to just the origin, rather than sending the whole path and query as a referrer.

Could you also add tests for the cases I mentioned?

Reply:
> urlForBindings looks a little strange here. This is not the bindings.

Thanks, yeah, changed to just `url()`

> We're also missing a call to strippedForUseAsReferrer

OK, added

> Also, I think we may want to restrict this to just the origin, rather than sending the whole path and query as a referrer.

Gecko and Blink send the whole path and query — so in order to have interop/compat with their existing behavior, the change here matches that.

But if we decide it’s desirable to not send the whole path and query, then I could switch the implementation to doing that — and then, I guess, raise Gecko and Blink bugs to try to get them to change (again, with the goal still being to get interop on this).

> Could you also add tests for the cases I mentioned?

See my previous comment — I added more tests (and the added tests show that Gecko and Blink are using the parent frame URL to set the Referer, not the “main” frame URL).